src/enzyme_ad/jax/Passes/RaiseTritonCustomCallPass.cpp

```diff
@@ -0,0 +1,173 @@
+//
+// This file implements a pass to raise stablehlo triton custom calls to the
+// triton ext dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include <zlib.h>
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Transforms/WalkPatternRewriteDriver.h"
+
+#include "src/enzyme_ad/jax/Passes/Passes.h"
+
+#include "src/enzyme_ad/jax/Dialect/TritonExt/Dialect.h"
+#include "src/enzyme_ad/jax/Utils.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+
+#include "jaxlib/gpu/triton.pb.h"
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_RAISETRITONCUSTOMCALLPASS
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::enzymexla;
+using namespace stablehlo;
+
+namespace {
+
+struct RaiseTritonCustomCallPattern final
+    : public OpRewritePattern<stablehlo::CustomCallOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(stablehlo::CustomCallOp callOp,
+                                PatternRewriter &rewriter) const {
+    auto callTargetName = callOp.getCallTargetName();
+    if (callTargetName != "triton_kernel_call")
+      return failure();
+
+    auto configAttr = callOp.getBackendConfig();
+    if (!configAttr)
+      return failure();
+
+    auto configStrAttr = dyn_cast<StringAttr>(*configAttr);
+    if (!configStrAttr)
+      return failure();
+
+    std::string config(configStrAttr.getValue());
+    uLongf destlen = config.size() * 5;
+    std::string data;
+
+    while (true) {
+      data.resize(destlen);
+      int ret =
+          uncompress(reinterpret_cast<Bytef *>(data.data()), &destlen,
+                     reinterpret_cast<Bytef *>(config.data()), config.size());
+      if (ret == Z_OK) {
+        data.resize(destlen);
+        break;
+      } else if (ret == Z_BUF_ERROR) {
+        destlen *= 2;
+      } else {
+        // cannot decompress backend config
+        return failure();
+      }
+    }
+
+    jax_triton::TritonAnyKernelCall proto;
+    if (!proto.ParseFromString(data))
+      return failure();
+
+    // TODO: support autotuned kernel calls
+    if (proto.has_autotuned_kernel_call())
+      return failure();
+
+    auto kCall = proto.kernel_call();
+    OwningOpRef<ModuleOp> mir = mlir::parseSourceString<ModuleOp>(
+        kCall.kernel().ttir(), callOp.getContext());
+    if (!mir)
+      return failure();
+
+    SymbolTable symTable(SymbolTable::getNearestSymbolTable(callOp));
+
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToEnd(&symTable.getOp()->getRegion(0).front());
+
+    auto innerMod = cast<mlir::ModuleOp>(mir.release().getOperation());
+    auto outerMod = enzymexla::triton_ext::TritonModuleOp::create(
+        rewriter, callOp.getLoc(), "triton_module");
```

Comment:
we might want to make sure there isn't a name clash when we raise multiple custom calls

Reply:
`symTable.insert` below will rename to a unique name like `triton_module_0` if it already exists.